### ISSUE TITLE
Cast to geo_shape not working on generated column

### ISSUE THREAD

**Issue by hlcianfagna** [crate/crate#16355]
### CrateDB version

5.7.3 , 5.8.0 , and nightly-5.9.0-2024-07-22-00-02

### CrateDB setup information

Tested both single-node and a 3-nodes cluster

### Problem description

I need to extract a `geo_shape` definition from another field and index it using bkdtree , using a generated column does not seem possible when a casting to `geo_shape` is involved.

### Steps to Reproduce

```sql
CREATE TABLE test (
	a OBJECT,
	b GENERATED ALWAYS AS a::geo_shape
);

SELECT {"coordinates"= [50.0,50.0],"type"= 'Point'}::geo_shape;
--> works

INSERT INTO test (a)
VALUES({"coordinates"= [50.0,50.0],"type"= 'Point'});
```

### Actual Result

`INSERT OK, 0 records affected`

Reproing this, at one point, I also came across the following stack trace, which I could not get a second time, but pasting it here as it may give some relevant clue:

```
org.elasticsearch.common.io.stream.NotSerializableExceptionWrapper: class_cast_exception: class io.crate.metadata.GeneratedReference cannot be cast to class io.crate.metadata.GeoReference (io.crate.metadata.GeneratedReference and io.crate.metadata.GeoReference are in unnamed module of loader 'app')
	at io.crate.execution.dml.GeoShapeIndexer.<init>(GeoShapeIndexer.java:64)
	at io.crate.types.GeoShapeType$1.valueIndexer(GeoShapeType.java:57)
	at io.crate.types.DataType.valueIndexer(DataType.java:255)
	at io.crate.execution.dml.Indexer.<init>(Indexer.java:514)
	at io.crate.execution.dml.upsert.TransportShardUpsertAction.processRequestItems(TransportShardUpsertAction.java:168)
	at io.crate.execution.dml.upsert.TransportShardUpsertAction.processRequestItems(TransportShardUpsertAction.java:93)
	at io.crate.execution.dml.TransportShardAction$1.call(TransportShardAction.java:105)
	at io.crate.execution.dml.TransportShardAction$1.call(TransportShardAction.java:102)
	at io.crate.execution.dml.TransportShardAction.wrapOperationInKillable(TransportShardAction.java:130)
	at io.crate.execution.dml.TransportShardAction.lambda$shardOperationOnPrimary$0(TransportShardAction.java:108)
	at org.elasticsearch.action.ActionListener.completeWith(ActionListener.java:239)
	at io.crate.execution.dml.TransportShardAction.shardOperationOnPrimary(TransportShardAction.java:98)
	at io.crate.execution.dml.TransportShardAction.shardOperationOnPrimary(TransportShardAction.java:53)
	at org.elasticsearch.action.support.replication.TransportReplicationAction$PrimaryShardReference.perform(TransportReplicationAction.java:926)
	at org.elasticsearch.action.support.replication.ReplicationOperation.execute(ReplicationOperation.java:124)
	at org.elasticsearch.action.support.replication.TransportReplicationAction$AsyncPrimaryAction.runWithPrimaryShardReference(TransportReplicationAction.java:427)
	at org.elasticsearch.action.support.replication.TransportReplicationAction$AsyncPrimaryAction.lambda$doRun$0(TransportReplicationAction.java:346)
	at org.elasticsearch.action.ActionListener$2.onResponse(ActionListener.java:98)
	at org.elasticsearch.index.shard.IndexShard.lambda$wrapPrimaryOperationPermitListener$21(IndexShard.java:2721)
	at org.elasticsearch.action.ActionListener$3.onResponse(ActionListener.java:126)
	at org.elasticsearch.index.shard.IndexShardOperationPermits.acquire(IndexShardOperationPermits.java:292)
	at org.elasticsearch.index.shard.IndexShardOperationPermits.acquire(IndexShardOperationPermits.java:242)
	at org.elasticsearch.index.shard.IndexShard.acquirePrimaryOperationPermit(IndexShard.java:2695)
	at org.elasticsearch.action.support.replication.TransportReplicationAction.acquirePrimaryOperationPermit(TransportReplicationAction.java:867)
	at org.elasticsearch.action.support.replication.TransportReplicationAction$AsyncPrimaryAction.doRun(TransportReplicationAction.java:342)
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37)
	at org.elasticsearch.action.support.replication.TransportReplicationAction.handlePrimaryRequest(TransportReplicationAction.java:308)
	at org.elasticsearch.transport.RequestHandlerRegistry.processMessageReceived(RequestHandlerRegistry.java:59)
	at org.elasticsearch.transport.TransportService$7.doRun(TransportService.java:682)
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.lang.Thread.run(Thread.java:1583)
```

### Expected Result

1 record inserted

**Comment by hlcianfagna** [comment]
Using an UDF to workaround this  does not seem to be allowed either.

**Comment by mfussenegger** [comment]
Thanks for the report.

This should be fixed with https://github.com/crate/crate/pull/16378 which will be released with 5.7.4 and 5.8.1
